Server db 1

oleksandrkravchuk87 opened the pull request:
comments about structures location taken into account, added functions for tables filling

oleksandrkravchuk87 committed:
database credentials are stored in config.yaml

oleksandrkravchuk87 committed:
init() function replaced with InitPostgresConfig()

AndriiPryslupskyi committed:
added server

oleksandrkravchuk87 committed:
Merge branch 'server' of https://github.com/ASeegull/SmartFridge

oleksandrkravchuk87 committed:
added postgres functions

oleksandrkravchuk87 committed:
adding database fill functions

oleksandrkravchuk87 committed:
adding database fill functions

oleksandrkravchuk87 committed:
changed Recipes function

oleksandrkravchuk87 committed:
changed Recipes function

oleksandrkravchuk87 committed:
changed Recipes function

oleksandrkravchuk87 committed:
added gorm ignote to Ingr field

oleksandrkravchuk87 committed:
added GetAllProductsID() function

8tomat8 commented on cmd/server/main.go:
```diff
@@ -1,10 +1,16 @@
-	fmt.Println(examplePackage.Func("server"))
+	host, port := server.GetAddr()
+	router := server.NewRouter()
+	fmt.Printf("Server started with address %s:%s\n", host, port)
```
Please use log packages for logging

8tomat8 commented on server/database/mongo.go:
```diff
@@ -0,0 +1,107 @@
+
+	yamlFile, err := ioutil.ReadFile("../SmartFridge/mongoConfig.yaml")
+	if err != nil {
+		log.WithFields(log.Fields{"error: ": err}).Info("Cannot read ../SmartFridge/mongoConfig.yaml")
```
DRY

8tomat8 commented on server/database/mongo.go:
```diff
@@ -0,0 +1,107 @@
+	}
+	err = yaml.Unmarshal(yamlFile, config)
+	if err != nil {
+		log.WithFields(log.Fields{"error: ": err}).Info("Unmarshal yamlFile error")
```
You have to error should not be used as field. Fields are created to store metadata

8tomat8 commented on server/database/mongo.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+func init() {
+	mongoConfig.getConf()
```
Try to avoid `init()` func

8tomat8 commented on server/database/mongo.go:
```diff
@@ -0,0 +1,107 @@
+func getSession() (*mgo.Session, error) {
+	var mongoURI = mongoConfig.MongoURI
+
+	dialInfo, _ := mgo.ParseURL(mongoURI)
```
Do not skip error checks

8tomat8 commented on server/database/mongo.go:
```diff
@@ -0,0 +1,107 @@
+
+	session, err := getSession()
+	if err != nil {
+		log.WithFields(log.Fields{"error: ": err}).Info("Cannot create session to mongoDB")
```
You have to error should not be used as field. Fields are created to store metadata

8tomat8 commented on server/database/mongo.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+//SaveState saves state from agent
+func SaveState(agentInfo *FoodAgent) {
```
Please return errors from functions if it possible.

8tomat8 commented on server/database/mongo.go:
```diff
@@ -0,0 +1,107 @@
+	session, err := getSession()
+	if err != nil {
+		log.WithFields(log.Fields{"error: ": err}).Info("Cannot create session to mongoDB")
+		return []FoodInfo{}
```
structs, slices, maps and interfaces has zero type - `nil`. So you have not to create empty slice to return, you can use `nil` instead of it

8tomat8 commented on server/database/mongo.go:
```diff
@@ -0,0 +1,107 @@
+	defer session.Close()
+	c := session.DB(mongoConfig.Database).C(mongoConfig.Table)
+
+	var foods []FoodInfo
```
`foods := make([]FoodInfo, 0, len(containersID))`

8tomat8 commented on server/database/mongo.go:
```diff
@@ -0,0 +1,107 @@
+	for _, value := range containersID {
+		var agent FoodAgent
+
+		if err := c.Find(bson.M{"token": value}).One(&agent); err == nil {
```
Handle error if it not nil. For example it could be connection error and you will have to reconnect

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+		dbUser, dbPassword, dbName)
+	db, err := gorm.Open("postgres", dbinfo)
+	if err != nil {
+		return errors.New("db connection error")
```
Do not omit `err`

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+	}
+	defer db.Close()
+	user := User{}
+	db.Where("login = ?", login).Find(&user)
```
`err`

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+	user.Login = login
+	user.Password = passHash
+	rows := db.Create(&user).RowsAffected
+	if !(rows == int64(1)) {
```
`if rows != int64(1) {`

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+	rows := db.Create(&user).RowsAffected
+	if !(rows == int64(1)) {
+		return errors.New("user not added")
+	} else {
```
`else` is extra

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+	defer db.Close()
+	user := User{}
+	db.Where("login = ?", login).Find(&user)
+	if user.ID == "" {
```
Replace it with `switch`

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+	db.Where("login = ?", login).Find(&user)
+	if user.ID == "" {
+		return errors.New("login not found")
+	} else if !(strings.TrimRight(pass, "\n") == user.Password) {
```
`!=`

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+	}
+	defer db.Close()
+	user := User{}
+	db.Where("login = ?", login).Find(&user)
```
`err`

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+	}
+	defer db.Close()
+	agent := Agent{}
+	db.Where("agents.id = ? AND agents.user_id = ?", idAgent, idUser).Find(&agent)
```
`err`

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+	db.Where("name LIKE ?", productName).First(&product)
+	if product.Name == "" {
+		return time.Time{}, errors.New("product not found")
+	} else {
```
extra `else`

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+
+//GetAllAgentsIDForClient returns all agent for clientID as a sice of string
+func GetAllAgentsIDForClient(userID string) ([]string, error) {
+	dbinfo := fmt.Sprintf("user=%s password=%s dbname=%s sslmode=disable",
```
Please move this common `dbinfo` to some metadata. You don't have to compose it each time.

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+	}
+	defer db.Close()
+	var productID int
+	productIDs := make([]int, 0, 100)
```
Why 100?

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+		return nil, errors.New("db connection error")
+	}
+	defer db.Close()
+	productNameSlice := make([]string, 0, 7)
```
Why 7?

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+}
+
+//Recipes(foodInfoSlice []FoodInfo) takes the slice of FoodInfo strucktures, representing all available products in all agents and return all recepies, which can be offered as a JSON
+func Recipes(foodInfoSlice []FoodInfo) ([]byte, error) {
```
Is this method okey for gometalinter?

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+	var name, unit string
+	var amount int
+OUTER:
+	for k := len(recipes) - 1; k >= 0; k-- {
```
`for k, recipe := range recipes {`

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+			if contains(productNameSlice, name) && amount <= productMap[name] {
+				recipes[k].Ingred = append(recipes[k].Ingred, strconv.Itoa(amount), unit, name)
+			} else {
+				recipes = append(recipes[:k], recipes[k+1:]...)
```
It is a bad practice to change slice that you are using for iteration.

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+	if err != nil {
+		return nil, errors.New("marshaling error")
+	}
+	return allRecipies, nil
```
`return json.Marshal(recipes)`

8tomat8 commented on server/database/postgres.go:
```diff
@@ -0,0 +1,368 @@
+	r = Recepie{ID: 3, Name: "Soup", Description: "...", CoockingTimeMin: 35, Complexity: "easy"}
+	db.Create(&r)
+
+	i := Ingridient{RecipeID: 1, ProductID: 1, Amount: 300}
```
Why don't you use `for` ?))) You are engineer, be lazy - automate it!

8tomat8 commented on server/fridgeServer.go:
```diff
@@ -0,0 +1,175 @@
+	router.HandleFunc("/agent", agentAuthentication).Methods("POST")
+	router.HandleFunc("/agent", createWS).Methods("GET")
+
+	router.HandleFunc("/client/allRecipes", getRecipes).Methods("GET")
```
You can group handlers to avoid copying of `/client`

8tomat8 commented on server/fridgeServer.go:
```diff
@@ -0,0 +1,175 @@
+		return
+	}
+
+	clientID := string(body[1 : len(body)-2])
```
WTF?
Please use `json`

8tomat8 commented on server/fridgeServer.go:
```diff
@@ -0,0 +1,175 @@
+		foods = database.GetFoodsInFridge(IDs)
+	}
+
+	json.NewEncoder(w).Encode(foods)
```
Where is error handling?

8tomat8 commented on server/fridgeServer.go:
```diff
@@ -0,0 +1,175 @@
+
+func agentAuthentication(w http.ResponseWriter, r *http.Request) {
+	tokenExample := getHash("userID+agentID+productName")
+	json.NewEncoder(w).Encode(tokenExample)
```
Error handling

oleksandrkravchuk87 committed:
added error handling

oleksandrkravchuk87 committed:
added error handling

oleksandrkravchuk87 committed:
added gorm error handling
simplified AllRecepies function
changer Recepies function (used copy of the slice)
"magic numbers" explained

oleksandrkravchuk87 committed:
init() function replaced with InitPostgresConfig()

oleksandrkravchuk87 committed:
init() function replaced with InitPostgresConfig()

oleksandrkravchuk87 committed:
changed the struckture

oleksandrkravchuk87 committed:
updated glide config

oleksandrkravchuk87 committed:
added role field to user structure
